<?xml version="1.0"?>
<robot name="servbot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Materials -->
  <material name="white"><color rgba="1 1 1 1" /></material>
  <material name="orange"><color rgba="1 0.3 0.1 1"/></material>
  <material name="blue"><color rgba="0.2 0.2 1 1"/></material>
  <material name="black"><color rgba="0 0 0 1"/></material>
  <material name="red"><color rgba="1 0 0 1"/></material>

  <!-- Base link -->
  <link name="base_link"/>

  <!-- Base footprint link -->
  <link name="base_footprint"/>

  <joint name="base_footprint_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_footprint"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <!-- Left wheel -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel_link"/>
    <origin xyz="0 -0.325 0" rpy="1.57 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <link name="left_wheel_link">
    <visual>
      <geometry><cylinder radius="0.05" length="0.023"/></geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="1e-4" iyy="1e-4" izz="1e-4" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- Right wheel -->
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel_link"/>
    <origin xyz="0 0.325 0" rpy="-1.57 0 0"/>
    <axis xyz="0 0 -1"/>
  </joint>

  <link name="right_wheel_link">
    <visual>
      <geometry><cylinder radius="0.05" length="0.023"/></geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="1e-4" iyy="1e-4" izz="1e-4" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- Front caster -->
  <joint name="front_caster_joint" type="fixed">
    <parent link="base_link"/>
    <child link="front_caster_link"/>
    <origin xyz="0.45 -0.105 0" rpy="0 0 0"/>
  </joint>

  <link name="front_caster_link">
    <visual>
      <geometry><sphere radius="0.05"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><sphere radius="0.05"/></geometry>
    </collision>
    <inertial>
      <mass value="0.01"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-5" iyy="1e-5" izz="1e-5" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- Rear caster -->
  <joint name="rear_caster_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rear_caster_link"/>
    <origin xyz="-0.45 -0.105 0" rpy="0 0 0"/>
  </joint>

  <link name="rear_caster_link">
    <visual>
      <geometry><sphere radius="0.05"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><sphere radius="0.05"/></geometry>
    </collision>
    <inertial>
      <mass value="0.01"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-5" iyy="1e-5" izz="1e-5" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- Chassis -->
  <link name="chassis_link">
    <visual>
      <geometry><box size="0.9 0.65 0.1"/></geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <mass value="1.0"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-2" iyy="1e-2" izz="1e-2" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <joint name="chassis_joint" type="fixed">
    <parent link="base_link"/>
    <child link="chassis_link"/>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
  </joint>

  <!-- IMU -->

  <joint name="imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="imu_link"/>
    <origin xyz="0.02 0.02 0.1" rpy="0 0 0"/>
  </joint>

  <link name="imu_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry><sphere radius="0.02"/></geometry>
      <material name="blue"/>
    </visual>
  </link>


  <!-- LIDAR -->

  <joint name="rplidar_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rplidar_link"/>
    <origin xyz="-0.15 -0.165 0.235" rpy="0 0 0"/>
  </joint>

  <link name="rplidar_link">
    <visual>
      <origin xyz=" 0 0 0" rpy="0 0 0"/>
      <geometry><cylinder radius="0.05" length="0.02"/></geometry>
      <material name="black"/>
    </visual>
  </link>

  <!-- ROS 2 Control -->
  <ros2_control name="DiffDriveHardware" type="system">
    <hardware>
      <plugin>diffdrive_hardware/DiffDriveHardware</plugin>
      <param name="serial_port">/dev/ttyACM0</param>
      <param name="baud_rate">115200</param>
      <param name="wheel_radius">0.05</param>
      <param name="wheel_separation">0.65</param>
      <param name="ticks_per_revolution">2800</param>
    </hardware>
    <joint name="left_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="right_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <sensor name="imu_sensor">
      <state_interface name="linear_acceleration.x"/>
      <state_interface name="linear_acceleration.y"/>
      <state_interface name="linear_acceleration.z"/>
      <state_interface name="angular_velocity.x"/>
      <state_interface name="angular_velocity.y"/>
      <state_interface name="angular_velocity.z"/>
    </sensor>
  </ros2_control>

</robot>